pymilvus/client/search_iterator.py

```diff
@@ -0,0 +1,152 @@
+from pymilvus.orm.iterator import SearchPage, fall_back_to_latest_session_ts
+
+logger = logging.getLogger(__name__)
+
+
+class SearchIteratorV2:
+    _NOT_SUPPORT_V2_MSG = """
+    The server does not support Search Iterator V2.
+    Please upgrade your Milvus server, or create a search_iterator (v1) instead.
+    """
+
+    # for compatibility, save the first result during init
+    _saved_first_res = None
+    _is_saved = False
+    # for compatibility, track the number of total results left
+    _left_res_cnt = None
+
+    def __init__(
+        self,
+        connection: Connections,
+        collection_name: str,
+        data: Union[List, utils.SparseMatrixInputType],
+        batch_size: int = 1000,
+        filter: Optional[str] = None,
+        output_fields: Optional[List[str]] = None,
+        search_params: Optional[Dict] = None,
+        timeout: Optional[float] = None,
+        partition_names: Optional[List[str]] = None,
+        anns_field: Optional[str] = None,
+        round_decimal: Optional[int] = -1,
+        **kwargs,
+    ):
+        self._check_params(batch_size, data, kwargs)
+
+        # for compatibility, delete limit from incoming
+        if MILVUS_LIMIT in kwargs:
+            self._left_res_cnt = kwargs[MILVUS_LIMIT]
+            del kwargs[MILVUS_LIMIT]
+
+        self._conn = connection
+        self._params = {
+            "collection_name": collection_name,
+            "data": data,
+            "anns_field": anns_field,
+            "param": deepcopy(search_params),
+            "limit": batch_size,
+            "expression": filter,
+            "partition_names": partition_names,
+            "output_fields": output_fields,
+            "timeout": timeout,
+            "round_decimal": round_decimal,
+            ITERATOR_FIELD: True,
+            ITER_SEARCH_V2_KEY: True,
+            ITER_SEARCH_BATCH_SIZE_KEY: batch_size,
+            GUARANTEE_TIMESTAMP: 0,
+            **kwargs,
+        }
+        # this raises MilvusException if the server does not support V2
+        self._saved_first_res = self.next()
+        self._is_saved = True
+
+    def next(self):
+        # for compatibility
+        if self._is_saved:
+            self._is_saved = False
+            return self._saved_first_res
+        self._saved_first_res = None
+        if self._left_res_cnt is not None and self._left_res_cnt <= 0:
+            return SearchPage(None)
+
+        res = self._conn.search(**self._params)
+        iter_info = res.get_search_iterator_v2_results_info()
+        self._params[ITER_SEARCH_LAST_BOUND_KEY] = iter_info.last_bound
+
+        # patch token and guarantee timestamp for the first next() call
+        if ITER_SEARCH_ID_KEY not in self._params:
+            if iter_info.token is not None and iter_info.token != "":
+                self._params[ITER_SEARCH_ID_KEY] = iter_info.token
+            else:
+                raise MilvusException(message=self.NOT_SUPPORT_V2_MSG)
```

Comment:
Please define a new Exception type like `ServerVersionIncompatibleException` with error msg NOT_SUPPORT_V2_MSG, so that we can cache this exception and avoid str comparison
```python
class ServerVersionIncompatibleException(MilvusException):
    pass

raise ServerVersionIncompatibleException


try:
    pass
except ServerVersionIncompatibleException as e:
   # v1 logic
else:
    # v2 logic

```